### ISSUE TITLE
MPI problem in multi-node jobs

### ISSUE THREAD

**Issue by sahmes** [amusecode/amuse#507]
I have a problem that occurs _usually_, but sometimes randomly it doesn't, when submitting a job to a cluster that has the SLURM workload manager and OpenMPI 3.1.2. The error message looks something like that:

    [node3:20840] [[40273,2],6] selected pml cm, but peer [[40273,1],0] on node3 selected pml ob1
    [node3:20835] [[40273,2],1] selected pml cm, but peer [[40273,1],0] on node3 selected pml ob1
    [node3:20836] [[40273,2],2] selected pml cm, but peer [[40273,1],0] on node3 selected pml ob1
    [node3:20835] [[40273,2],1] ORTE_ERROR_LOG: Unreachable in file dpm/dpm.c at line 433
    [node3:20838] [[40273,2],4] selected pml cm, but peer [[40273,1],0] on node3 selected pml ob1
    [node3:20838] [[40273,2],4] ORTE_ERROR_LOG: Unreachable in file dpm/dpm.c at line 433
    [node3:20836] [[40273,2],2] ORTE_ERROR_LOG: Unreachable in file dpm/dpm.c at line 433
    [node3:20837] [[40273,2],3] selected pml cm, but peer [[40273,1],0] on node3 selected pml ob1
    [node3:20840] [[40273,2],6] ORTE_ERROR_LOG: Unreachable in file dpm/dpm.c at line 433
    [node3:20837] [[40273,2],3] ORTE_ERROR_LOG: Unreachable in file dpm/dpm.c at line 433
    --------------------------------------------------------------------------
    It looks like MPI_INIT failed for some reason; your parallel process is
    likely to abort.  There are many reasons that a parallel process can
    fail during MPI_INIT; some of which are due to configuration or environment
    problems.  This failure appears to be an internal failure; here's some
    additional information (which may only be relevant to an Open MPI
    developer):

    ompi_dpm_dyn_init() failed
    --> Returned "Unreachable" (-12) instead of "Success" (0)
    --------------------------------------------------------------------------
    [node3:20839] [[40273,2],5] selected pml cm, but peer [[40273,1],0] on node3 selected pml ob1
    [node3:20839] [[40273,2],5] ORTE_ERROR_LOG: Unreachable in file dpm/dpm.c at line 433
    [node3:20840] *** An error occurred in MPI_Init_thread
    [node3:20840] *** reported by process [2639331330,6]
    [node3:20840] *** on a NULL communicator
    [node3:20840] *** Unknown error
    [node3:20840] *** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
    [node3:20840] ***    and potentially your MPI job)
    [node8:165900] [[40273,2],7] selected pml cm, but peer [[40273,1],0] on unknown selected pml ob1
    [node8:165900] [[40273,2],7] ORTE_ERROR_LOG: Unreachable in file dpm/dpm.c at line 433
    [node3:20801] 6 more processes have sent help message help-mpi-runtime.txt / mpi_init:startup:internal-failure
    [node3:20801] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
    [node3:20801] 5 more processes have sent help message help-mpi-errors.txt / mpi_errors_are_fatal unknown handle

In this example I tried to get two nodes, each with four GPU-using tasks (in the Python script I created a ph4 instance with 8 workers). This was my SLURM submission script:

    #!/bin/bash
    #SBATCH --gres=gpu:4
    #SBATCH --nodes=2
    #SBATCH --ntasks=8

    module load nixpkgs/16.09 gcc/7.3.0 cuda/10.0.130 openmpi/3.1.2
    export LD_PRELOAD=/cvmfs/soft.computecanada.ca/easybuild/software/2017/avx2/CUDA/gcc7.3/cuda10.0/openmpi/3.1.2/lib/libmpi.so
    export OMPI_MCA_rmaps_base_oversubscribe=yes
    mpiexec -n 1 python amuse_test.py

The setting of the strange environment variables `LD_PRELOAD` and `OMPI_MCA_rmaps_base_oversubscribe` turned out to be necessary, otherwise it doesn't even work on a single node. The latter environment variable is necessary because if I don't set it, I get an mpi4py error `MPI_ERR_SPAWN: could not spawn processes`, but I don't have full understanding of what the  repercussions of setting it to "yes" are, just that it prevents this error and AMUSE codes runs fine on a single node. Maybe some manipulation of other MCA parameters could solve my multi-node problem, but my attempts so far have been futile...

**Comment by ipelupessy** [comment]
The oversubscribe option is sometimes necessary with mpi (it means more processes than cpus are allowed)..so thats ok. Why do you need the LD_PRELOAD? anyway:
 Do you know if you encounter the same problem with other versions of openMPI and/or MPICH? Does the normal MPI hello world work? There is also a simple hello world example for spawning processes:
http://mpi.deino.net/mpi_functions/MPI_Comm_spawn.html
or
https://www.mcs.anl.gov/research/projects/mpi/mpi-standard/mpi-report-2.0/node98.htm

**Comment by sahmes** [comment]
The `LD_PRELOAD` variable is needed to fix yet another error

    ompi_mpi_init: ompi_rte_init failed
    --> Returned "(null)" (-43) instead of "Success" (0)
    --------------------------------------------------------------------------
    *** An error occurred in MPI_Init_thread
    *** on a NULL communicator

I can't remember where online I found this `LD_PRELOAD` trick, I don't fully understand it, but it works (for a single node with an arbitrary number of workers).

On this cluster I have no problem running other MPI codes on multiple nodes (without setting any special environment variables, the submission scripts are usually trivial), I haven't tried anything else with mpi4py other than AMUSE.

Unfortunately I don't have the luxury of testing multiple MPI implementations or versions, for some reason only one MPI module on the cluster is compatible with CUDA, although I suppose I could try to compile in my home folder.

**Comment by ipelupessy** [comment]
are you sure that the mpi4py is build against  the openmpi version used? 
It seems could be that the LD_PRELOAD is necessary to get mpi4py use the version indicated; have you tried setting the LD_LIBRARY_PATH instead?
you can try running an python script with just:
```
mpiexec python -c "from mpi4py import MPI"
```
and maybe omit the mpiexec?

**Comment by sahmes** [comment]
So after rebuilding mpi4py (previously I just installed via pip), it seems that setting `LD_PRELOAD` isn't necessary anymore.

Now when executing `mpiexec python -c "from mpi4py import MPI"` from a compute node with a GPU I get this message:

    node7.172921psmi_cuda_initialize: Unable to find active CUDA context

It's the same if I omit the `mpiexec` and just run python. Curiously this does not seem to be an error, or at least not a fatal one. When I run my AMUSE script (that uses a gpu worker of ph4, just one worker in this case) I get the same message but the code is actually running and evolving the model successfully (I still have to set `OMPI_MCA_rmaps_base_oversubscribe=yes` for the AMUSE script to work). If I set `LD_PRELOAD` as before, I don't get this or any warning, and things still work fine (both importing mpi4py.MPI and running the AMUSE sctipt).

However reinstalling mpi4py did not change the multi-node problem, I still get the same type of error message regarding PML incompatibility.
